arc/src/main.c

```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
     ipm_send_reply(msg);
 }
 
+#ifndef DEBUG_BUILD
+static inline int sensor_value_snprintf(char *buf, size_t len,
+                                        const struct sensor_value *val)
+{
+    int32_t val1, val2;
+
+    switch (val->type) {
+    case SENSOR_VALUE_TYPE_INT:
+        return snprintf(buf, len, "%d", val->val1);
+    case SENSOR_VALUE_TYPE_INT_PLUS_MICRO:
+        if (val->val2 == 0) {
+            return snprintf(buf, len, "%d", val->val1);
+        }
+
+        /* normalize value */
+        if (val->val1 < 0 && val->val2 > 0) {
+            val1 = val->val1 + 1;
+            val2 = val->val2 - 1000000;
+        } else {
+            val1 = val->val1;
+            val2 = val->val2;
+        }
+
+        /* print value to buffer */
+        if (val1 > 0 || (val1 == 0 && val2 > 0)) {
+                return snprintf(buf, len, "%d.%06d", val1, val2);
+        } else if (val1 == 0 && val2 < 0) {
+                return snprintf(buf, len, "-0.%06d", -val2);
+        } else {
+                return snprintf(buf, len, "%d.%06d", val1, -val2);
+        }
+    case SENSOR_VALUE_TYPE_DOUBLE:
+        return snprintf(buf, len, "%f", val->dval);
+    default:
+        return 0;
+    }
+}
+#endif
+
+static void send_sensor_data(enum sensor_channel channel,
+                             union sensor_reading reading)
+{
+    struct zjs_ipm_message msg;
+    msg.id = MSG_ID_SENSOR;
+    msg.type = TYPE_SENSOR_EVENT_READING_CHANGE;
+    msg.flags = 0;
+    msg.user_data = NULL;
+    msg.error_code = ERROR_IPM_NONE;
+    msg.data.sensor.channel = channel;
+    memcpy(&msg.data.sensor.reading, &reading, sizeof(union sensor_reading));
+    zjs_ipm_send(MSG_ID_SENSOR, &msg);
+}
+
+#define ABS(x) ((x) > 0) ? (x) : -(x)
+
+static double convert_sensor_value(const struct sensor_value *val)
+{
+    int32_t val1, val2;
+    double result = 0;
+
+    switch (val->type) {
+    case SENSOR_VALUE_TYPE_INT:
+        result = val->val1;
+        break;
+    case SENSOR_VALUE_TYPE_INT_PLUS_MICRO:
+        if (val->val2 == 0) {
+            result = (double)val->val1;
+            break;
+        }
+
+        /* normalize value */
+        if (val->val1 < 0 && val->val2 > 0) {
+            val1 = val->val1 + 1;
+            val2 = val->val2 - 1000000;
+        } else {
+            val1 = val->val1;
+            val2 = val->val2;
+        }
+
+        if (val1 > 0 || (val1 == 0 && val2 > 0)) {
+            result = val1 + (double)val2 * 0.000001;
+        } else if (val1 == 0 && val2 < 0) {
+            result = (double)val2 * (-0.000001);
+        } else {
+            result = val1 + (double)val2 * (-0.000001);
+        }
```

Comment:
What happens in this block if val1 > 0 and val2 < 0, doesn't it do the wrong thing?

It kind of seems like this logic could be simplified to be more obviously correct. I'm still not sure what it's doing but will wait for your response to this before digging deeper. :)